test/Microsoft.AspNet.Server.KestrelTests/EngineTests.cs

```diff
@@ -57,6 +57,20 @@ ILibraryManager LibraryManager
             }
         }
 
+        private async Task AppThatThrows(Frame frame)
+        {
+            // Anything added to the ResponseHeaders dictionary is ignored
+            frame.ResponseHeaders["Content-Length"] = new[] { "11" };
+            throw new Exception();
+        }
+
+        private async Task AppThatThrowsAfterWrite(Frame frame)
+        {
+            frame.ResponseHeaders["Content-Length"] = new[] { "11" };
+            await frame.ResponseBody.WriteAsync(Encoding.UTF8.GetBytes("Hello World"), 0, 11);
```

Comment:
Because you specified Content-Length and then wrote that many bytes, the response is technically finished after this line. You should have an additional test that throws after having written only half of the data (and/or a response without Content-Length).